include/deal.II/numerics/smoothness_estimator.h

```diff
@@ -0,0 +1,385 @@
+   * @f]
+   * for $j=0,..,p$. The rate of the decay $\sigma$ can be used to estimate the
+   * smoothness. For example, one strategy to implement hp-refinement
+   * criteria is to perform p-refinement if $\sigma>1$.
+   *
+   * Extension to higher dimension is done by performing the fit in each
+   * coordinate direction separately and then taking the lowest value of
+   * $\sigma$.
+   *
+   * For each input vector of degrees of freedom defined on a DoFHandler,
+   * this function returns a vector with as many elements as there are cells
+   * where each element contains $\exp(-\sigma)$, which is a so-called
+   * analyticity (see references below).
+   *
+   * @param [in] fe_series FESeries::Legendre object to calculate coefficients.
+   * This object needs to be initialized to have at least $p+1$ coefficients in
+   * each direction, where $p$ is the maximum polynomial degree to be used.
+   * @param [in] dof_hander An hp::DoFHandler
+   * @param [in] all_solutions A vector of pointers to the solution vectors
+   * @param [out] all_smoothness_indicators A vector of pointers to the
+   * smoothness indicators for each @p all_solutions.
+   * @param [in] coefficients_predicate A predicate to select Legendre
+   * coefficients $a_j \;\; j=0\dots p$ for linear regression in each coordinate
+   * direction. The user is responsible for updating the vector of `flags`
+   * provided to this function. Note that its size is $p+1$, where $p$ is the
+   * polynomial degree of the FE basis on a given element. Default
+   * implementation will use all Legendre coefficients in each coordinate
+   * direction, i.e. set all elements of the vector to `true`.
+   * @param [in] smallest_abs_coefficient The smallest absolute value of the
+   * coefficient to be used in linear regression in each coordinate direction.
+   * Note that Legendre coefficients of some functions may have a repeating
+   * pattern of zero coefficients (i.e. for functions that are locally symmetric
+   * or antisymmetric about the midpoint of the element in any coordinate
+   * direction). Thus this parameters allows to ingore small (in absolute value)
+   * coefficients within the linear regression fit. In case there are less than
+   * two non-zero coefficients for a coordinate direction, this direction will
+   * be skipped. If all coefficients are zero, the returned value for this cell
+   * will be zero (i.e. corresponding to the $\sigma=\infty$).
+   * @param [in] only_flagged_cells Smoothness indicators are usually used to
+   * decide whether to perform h- or p-adaptation. So in most cases, we only
+   * need to calculate those indicators on cells flagged for refinement or
+   * coarsening. This parameter controls whether this particular subset or all
+   * cells will be considered. By default, only flagged cells will be
+   * considered: Smoothness indicators will only be set on those vector entries
+   * of flagged cells; the others will be set to zero.
```

Comment:
I use `std::nan("")` instead of `0` now.

Reply:
I think you want to use `std::numeric_limits<double>::quiet_NaN()` or `numbers::signaling_nan()`.

Reply:
VisIt does not like it when `vtk` files contain these `nan` values.

Do you know a way to cope with that, or shall we make the user responsible for this? By default, all cells will be considered, so this will not come to the user as a surprise.

Reply:
I see. I had assumed that you compute this for all locally owned cells but that the vector also contains entries for ghost cells.

If Visit doesn't like nan values, does it copy with inf? Maybe `-inf` is a better indicator of invalid values then.

Reply:
We introduced a flag that will only compute smoothness indicators on cells which have a `refine` or `coarsen` flag assigned. This way, we have invalid values even on locally owned cells.

I will check how visit treats `inf` values. Does it make a difference to give it a negative sign?

One type of my indicators can be negative, too. (I'll take the difference between predicted and estimated errors.)

Reply:
Visit doesn't like `inf` either. Let's stick to `NaN` values, since all indicators will be calculated by default anyways users should be aware of them.

Reply:
Yes, let's stick with NaN. It's probably rare to want to output these indicators to begin with; if one wanted to do that, it would not be difficult to write a short loop that just replaces the NaNs by some other invalid value.